Cache limiter support attempt

pine3ree opened the pull request:
Inject cache headers based on cache_limiter value, while removing the automatic job from the php engine.

The php engine automatically sends cache-http-headers based on the values of `session.cache_limiter`, `session.cache_expire` ini/session_start settings and the script last mtime.

The idea is to steal the value of `session.cache_limiter` inside the constructor and set an empty cache_limiter when starting the session, thus disabling the automatic headers.

Then we recreate the logic of the zend-engine to build the headers ourselves and inject them into the response.

TODO/TOCHECK
- find a better/proper way to check the running script mtime
- decide if/when to inject the cache header: right now if a cache header is found we assume that it was added programmatically in an internal middleware layer, so we skip injecting any of them. What should we do: override-headers? add-(multi-value)-headers?

pine3ree committed:
Inject cache headers based on cache_limiter value

pine3ree committed:
pushing tests to upstream

pine3ree committed:
remove redundant assignments

pine3ree committed:
add space after not operator

pine3ree committed:
change generateCacheHeaders signature

pine3ree committed:
add return types + change getLastModified() signature to accept filename param

pine3ree committed:
added test for empty cache_limiter

pine3ree committed:
added last-modified test

pine3ree commented:
need to fix lowered coverage...

pine3ree committed:
added last-modified failure test

pine3ree committed:
added unsupported cache_limiter value test

pine3ree committed:
update getLastModified method: if is_file can do filemtime

pine3ree committed:
fix last coverage failure: condition is never met (cacheLimiter is checked before call)

pine3ree commented:
Fixes https://github.com/zendframework/zend-expressive-session-ext/issues/10

pine3ree committed:
add supported limiters property and de-uglify code

pine3ree committed:
fix cs space after not

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -25,6 +25,10 @@
+use function sprintf;
+use function gmdate;
+use function time;
+use function filemtime;
```
Import statements should be alphabetized (simplifies finding declarations, and reduces chances of duplication).

pine3ree replied:
sorry, that's ignorance on my part...I always used to set the order based on which func was called first

weierophinney replied:
No worries - just providing some guidance. :)

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -44,6 +48,29 @@ class PhpSessionPersistence implements SessionPersistenceInterface
+        'private_no_expire' => true,
+    ];
+
+    public const CACHE_PAST_DATE  = 'Thu, 19 Nov 1981 08:52:00 GMT';
```
Why this date in particular, and not the unix epoch?

pine3ree replied:
I was curious about that value as well :smile:
https://github.com/php/php-src/blob/master/ext/session/session.c#L1194

weierophinney replied:
Wow, that's super bizarre! I guess we should use the value for consistency with the engine... maybe throw a comment in indicating where we get the value from, though. :)

pine3ree replied:
Yeah, I used it for consistency....but I'm still wondering where (or when) it comes from....

weierophinney replied:
Probably some PHP contributor's birthday. :grinning:

pine3ree replied:
https://stackoverflow.com/questions/8194481/why-is-expires-1981 :)

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -66,7 +93,21 @@ public function persistSession(SessionInterface $session, ResponseInterface $res
+                }
+            }
+
+            return $response;
```
Invert the conditional so you can return early, combine conditions, and promote the body of the conditional:

```php
if (! $this->cacheLimiter || $this->responseAlreadyHasCacheHeaders($response)) {
    return $response;
}

$cacheHeaders = $this->generateCacheHeaders($this->cacheLimiter, $this->cacheExpire);
// ...
return $response;
```

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -105,4 +147,82 @@ private function generateSessionId() : string
+    /**
+     * Generate cache headers for a given session cache_limiter value.
+     * @param string $cacheLimiter
+     * @param int $cacheExpire
```
Parameter annotations are discouraged unless they add descriptions; these can be removed.

pine3ree replied:
do I keep the method's purpose explaination?

weierophinney replied:
Yes.

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -105,4 +147,82 @@ private function generateSessionId() : string
+        }
+
+        $maxAge       = 60 * $cacheExpire;
+        $lastModified = $this->getLastModified($_SERVER['SCRIPT_FILENAME'] ?? '');
```
We typically try not to access superglobals directly. Please pass this value to the method by using the following from within `initializeSessionFromRequest()`:

```php
$request->getServerParams()['SCRIPT_FILENAME']
```

Considering that this may not be set, you'll need to come up with a default value to use; in that case, I'd likely use `__FILE__`:

```php
$request->getServerParams()['SCRIPT_FILENAME'] ?? __FILE__
```

pine3ree replied:
Yeah, I thought about that too...I was actually waiting for your review. I actually used `__FILE__` in the test case

I'd have a few questions before applying the changes:

1. generateCacheHeaders(...) is privately used only with internal properties, should I remove the arguments and use them directly?

2. same question getLastModified, since the scriptFile will be saved into an internal property

3. Should I remove thi PR and re-add a 1-commit PR? (I started uploading to upstream while forgetting i had already opened the PR, I apologize for the number of commits)

weierophinney replied:
1. Sure, that should work fine.
2. Same.
3. No, I'm fine with multiple commits. If YOU are not, then run `git rebase -i` and squash some commits and force push back to your branch.

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -105,4 +147,82 @@ private function generateSessionId() : string
+
+    /**
+     * Return the Last-Modified header line based on script name mtime
+     * @return string
```
This parameter annotation is unnecessary and can be removed.

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -105,4 +147,82 @@ private function generateSessionId() : string
+    /**
+     * Check if the response already carries cache headers
+     * @param ResponseInterface $response
+     * @return bool
```
Both of these annotations may be removed, as they duplicate the typehints.

weierophinney commented on src/PhpSessionPersistence.php:
```diff
@@ -105,4 +147,82 @@ private function generateSessionId() : string
+    private function responseAlreadyHasCacheHeaders(ResponseInterface $response) : bool
+    {
+        return (
+               $response->hasHeader('Expires')
```
Reduce the indentation of this line; `||` operators should align with this line.

weierophinney commented on test/PhpSessionPersistenceTest.php:
```diff
@@ -23,6 +23,8 @@
 use function session_start;
 use function session_status;
+use function time;
+use function gmdate;
```
Alphabetize imports, please.

weierophinney commented on test/PhpSessionPersistenceTest.php:
```diff
@@ -159,4 +161,186 @@ public function testPersistSessionIfSessionHasContents()
+        $expires = strtotime($expires);
+
+        $this->assertTrue($expires >= $expiresMin);
+        $this->assertTrue($expires <= $expiresMax);
```
Please use `assertGreaterThanOrEqual()` and/or `assertLessThanOrEqual()`.

weierophinney commented on test/PhpSessionPersistenceTest.php:
```diff
@@ -159,4 +161,186 @@ public function testPersistSessionIfSessionHasContents()
+
+        // temporarily set SCRIPT_FILENAME to current file for testing
+        $scriptFilename = $_SERVER['SCRIPT_FILENAME'] ?? null;
+        $_SERVER['SCRIPT_FILENAME'] = __FILE__;
```
With the changes I suggested earlier around usage of superglobals, this test will need to change.

What I didn't realize is that you're accessing it during `persistSession()` as well. I'm wondering if the script filename should be memoized in the Session instance, and removed during persistence?

pine3ree replied:
yes, see my previous reply...I don't think we need to remove it though whit it being refreshed at every request...and now that I think about it...should we just memoize it once....the running script is not going to change even in asynchronous contexts?

weierophinney replied:
Sure.

pine3ree replied:
....mhmmm....but mocked requests could contain a different SCRIPT_FILENAME....

pine3ree commented:
..i forgot another question:
I used public constants for past date and http date format and reused them in the tests, but, aside from the tests, I don't they are useful outside the class

weierophinney commented:
> I used public constants for past date and http date format and reused them in the tests, but, aside form the tests, I don't they are useful outside the class

Using them in tests is public behavior, so you can mark them public if you want. Otherwise you'll need to use reflection in your tests to get at the values, which is generally not fun. :)

pine3ree committed:
Implement changes requested by the reviewer (good man)

pine3ree commented:
mhmm...I'm not sure why test are failing on travis, everything works locally...I'll check later....

pine3ree committed:
remove unused argument: changed signature

pine3ree commented:
Still no clue, locally all test pass and in a test app the appropriate headers are correctly sent.
Could it be a travic issue?
What I find weird is that even a simple unchanged test case like the `cache_limiter` === nocache is failing, even if the session config is memoized as before and the code generating that constant headers array is basically unchanged (just remove the arguments and use the internal properties directly)

pine3ree committed:
re-add return type anotation and re-trigger travis tests

pine3ree committed:
trigger travis tests

pine3ree committed:
remove redundant retun and trigger new travis test

pine3ree commented:
@weierophinney 

Travis build continue to fail, I believe this has smt to do with the merging of my previous (unrelated) merged PR, even if there is no report of conflicts. My same fork and another test branch forked from it are still passing tests successfully. I also see a strange decrease in coverage.

https://travis-ci.org/pine3ree/zend-expressive-session-ext/builds/371829367

any idea?
kind regards

PS
there is another small change that I believe could make sense.

https://github.com/pine3ree/zend-expressive-session-ext/blob/00e0f7c9a3af7c0cc8ae4a456e83438946c1388f/src/PhpSessionPersistence.php#L222

In the getLastModified() method  we may have a memoized scriptFile either from a request server param or from the current `__FILE__`. In cases when the request scritpFile is an empty string or representes a non existent file, we still end up with a false value. I believe we should set it to `__FILE__` for those cases as well inside the function.

In the end we should always get a value, but since gmdate could also return `false`, we should assume that value can be returned as well and leave the fals check when injecting the headers 

```php
    //....
    public function initializeSessionFromRequest(ServerRequestInterface $request) : SessionInterface
    {
         // comment note: we set it to __FILE__ only once inside getLastModified if necessary
        $this->scriptFile = $request->getServerParams()['SCRIPT_FILENAME'] ?? null;
        //....
    }
    //....
    private function getLastModified()
    {
        if (! $this->scriptFile || ! is_file($this->scriptFile)) {
            $this->scriptFile = __FILE__;
        }

        return gmdate(self::HTTP_DATE_FORMAT, filemtime($this->scriptFile));
    }
    //----
```

(edited)
The test cases were not compatible with the last merge, I fixed it. regards.

pine3ree committed:
fix test cases for compatibility with recently merged fix

pine3ree committed:
add and use helper methods to improve test cases readability

pine3ree commented:
Hello @weierophinney,
I rebased to the latest merge and squashed into fewer commits for a cleaner history into a new branch:

- https://github.com/pine3ree/zend-expressive-session-ext/tree/cache-limiter 
_(replacing current PR cache-limiter-support)_
- https://github.com/pine3ree/zend-expressive-session-ext/branches 
_(patch-1 and patch-2 are unrelated to this matter, already opened different PRs for them)_
- [travis tests](https://travis-ci.org/pine3ree/zend-expressive-session-ext)
- [compare + cleaner-history](https://github.com/zendframework/zend-expressive-session-ext/compare/master...pine3ree:cache-limiter)

When you review last commits/comments pls let me also know if you prefer a PR from the new branch (closing this one)

kind regards

weierophinney reviewed (approved):
Final changes look perfect, @pine3ree !

weierophinney commented:
Thanks, @pine3ree! Merged to develop for release with 1.1.0.
